### PR TITLE
Catch JSON deserialization exceptions and use them to return 400 errors

### PR DESCRIPTION
**Pull request by jeffw-wherethebitsroam**
This code adds the JSON deserialization exception to
GraphQLRequestDeserializationResult and sets IsSuccessful = false.
Also, unmatched content-type should return 415 error code: Unsupported Media Type

### CODE REVIEW

**Comment by jeffw-wherethebitsroam** [comment]
This addresses some of the issues from https://github.com/graphql-dotnet/server/issues/580

**Review by Shane32 — APPROVED** [review]
Looks good to me, but @sungam3r should review.

File: tests/Samples.Server.Tests/ResponseTests.cs
Comment (sungam3r) [inline tests/Samples.Server.Tests/ResponseTests.cs:101]: > // POST with unsupported mime type should be a bad request

Maybe change this comment ?

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:110]: Let's change `httpStatusCode` argument type from `WriteErrorResponseAsync` to `System.Net.HttpStatusCode.BadRequest` and pass enum so no comment is needed. Then cast argument to `int` inside.

File: src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs
Comment (sungam3r) [inline src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs:47]: I would not add try/catch inside each deserializer. Let's exception bubble up the stack and catch it in `GraphQLHttpMiddleware` so you don't need new `Exception` property.  Even further. Now I think that `IsSuccessful` property is redundant since we can just throw exception. In this case, API will become simpler - either good deserialized result or exception, no third option via `bool` flag.

**Review by sungam3r — CHANGES_REQUESTED** [review]
Need small changes in the API design. I'm OK to merge after that.

File: tests/Samples.Server.Tests/ResponseTests.cs
Comment (jeffw-wherethebitsroam) [reply tests/Samples.Server.Tests/ResponseTests.cs:98]: Yep, done.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (jeffw-wherethebitsroam) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs]: I've change the status code int to a HttpStatusCode.

File: src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs
Comment (jeffw-wherethebitsroam) [reply src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs:47]: The was my first thoughts as well, but _serializer.Deserialize will use either `Newtonsoft.Json` or `System.Text.Json`. Which means we would either have to:
1. Catch both `Newtonsoft.Json.JsonException` and `System.Text.Json.JsonException` in middleware, which makes for a very leaky abstraction.
2. Catch `Exception` and assume it was a a json parsing exception :(
3. Make a wrapper exception, e.g. `GraphQLRequestDeserializerJsonException` to be thrown be both deserializers.

**Comment by SK201** [comment]
Hello,
We are currently affected by this issue and would love to see the PR merged.

Thanks in advance !

**Comment by jeffw-wherethebitsroam** [comment]
I was waiting on feedback on how to propagate the exceptions and then forgot about this. I can update the PR :)

File: src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs
Comment (jeffw-wherethebitsroam) [reply src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs]: I have implemented 3. using a wrapper exception.

**Comment by Shane32** [comment]
Thanks @jeffw-wherethebitsroam !  Did the API surface change?  We might need the apiapprovals file updated.  Also it looks like it is failing the formatting check.

**Comment by Shane32** [comment]
@jeffw-wherethebitsroam I found the issue with the formatting tool; if you can merge master back into this branch, then it should run the CI tests properly.

**Comment by jeffw-wherethebitsroam** [comment]
@Shane32 I've merged master. The API surface did change (removed IsSuccessful from GraphQLRequestDeserializationResult, added GraphQLRequestDeserializationException), but the apiapprovals file is updated.

**Comment by jeffw-wherethebitsroam** [comment]
@Shane32 It looks like I need a maintainer to approve the workflows: "First-time contributors need a maintainer to approve running workflows"

**Comment by Shane32** [comment]
I think these changes look great!  However, it contains breaking changes in the API surface, and has breaking changes to the semantics of the current request parsing methods.  It really belongs in the next major version, I would think.

To make this a non-breaking change, we could retain the `IsSuccessful` property and add another `Exception` property of type `GraphQLRequestDeserializationException`, or `null` if no error.  This really would perform identically, even though it's not as clean of a design.  Then the decision would become if we should clean up the design for the next major version or not.

The remainder of the semantic changes, such as returning the proper status code upon failure, would only necessitate a bump to the minor version number.

There is one other alternative to getting these changes in the current version, and that would be to leave the `IsSuccessful` property there, but always return `true`, and mark it as obsolete.  The property would need to be monitored in the middleware in case someone is using a custom JSON transport that returns `false` for that property.  I don't really care for this idea.

Thoughts?

**Comment by jeffw-wherethebitsroam** [comment]
Ok, we're now pretty much back to the original version :) Its a bit uglier now since `IsSuccessful = false` and `Exception != null` both mean that the deserialization failed and either or both can be set. But it would be better to clean that up in the next major release.

**Review by Shane32 — APPROVED** [review]

**Comment by Shane32** [comment]
I just have so little time to spend on the next version, and there’s a whole lot of cleanup that could be done. Better to get this fix in the current version.

**Comment by sungam3r** [comment]
@jeffw-wherethebitsroam Thanks for this fix. Just reviewed it. LGTM, just a few cosmetic changes [here](#651). I had a long break from OSS this year but I returned. I think that when I read the remaining emails for this repo (~60 left), then I can release v5.1 including this fix (and others of course).

**Pull request merged**
